### PR TITLE
Added the closed_tabs button - work in progress

### PR DESCRIPTION
**Pull request by bjedrzejewski**
If you want to have the look already there so everyone can comment / suggest / fix

### CODE REVIEW

**Comment by owennw** [comment]
Should the button still be visible when the search tab is open?

**Comment by bjedrzejewski** [comment]
@owennw that was going to be my question. What do you think? Maybe I will make it fade?

**Comment by owennw** [comment]
The icon should remain, but clicking it should not reopen the favourites tab. Also FYI there is a design for the icon's placement in compact view (it's moved to the toolbar, left of the minimise icon) but you may not want to do that in this PR / you should check for merge issues first.

**Comment by owennw** [comment]
Having looked at the spec again we (I) may have misinterpreted the two icons. It looks like the blue one is for when a window has recently closed, and grey for when the closed windows have been seen, I think. I don't know if the blue one was also intended as a hover-state or not.

File: src/app.js [hunk 1]
```diff
@@ -15,7 +15,9 @@
         'openfin.star',
         'openfin.tearout',
         'openfin.minichart',
-        'openfin.scroll'
+        'openfin.scroll',
+        'openfin.closedwindows',
+        'openfin.closedcard'
```
Comment (owennw) [inline src/app.js]: I feel these should be camel-cased.

File: src/assets/styles/style.less [hunk 2]
```diff
@@ -3,6 +3,7 @@
 
 @import "variables.less";
 @import "../../main/main.less";
+@import "../../closedwindows/closedcard.less";
```
Comment (owennw) [inline src/assets/styles/style.less]: Similar camel-casing here

File: src/closedwindows/closedcard-controller.js [hunk 3]
```diff
@@ -0,0 +1,13 @@
+(function() {
+    'use strict';
+
+    class ClosedcardCtrl {
+        constructor() {
+
+        }
```
Comment (owennw) [inline src/closedwindows/closedcard-controller.js]: The constructor can be deleted.

File: src/app.js (hunk 1 above)
Comment (bjedrzejewski) [inline src/app.js]: minichart is not, I followed the convention

File: src/closedwindows/closedcard-controller.js [hunk 4]
```diff
@@ -0,0 +1,13 @@
...
+    }
+    ClosedcardCtrl.$inject = [];
+
+    angular.module('openfin.closedcard', [])
```
Comment (owennw) [inline src/closedwindows/closedcard-controller.js]: The injected parameters shouldn't be declared again here (they're declared in the app.js).

File: src/closedwindows/closedcard-controller.js (hunk 3 above)
Comment (bjedrzejewski) [inline src/closedwindows/closedcard-controller.js]: I will be adding it right back in the net work.

File: src/closedwindows/closedcard.html [hunk 5]
```diff
@@ -0,0 +1,7 @@
+<div class="closedcard">
+    <div class="closedtime">{{cwindow.closed | date:'yyyy-MM-dd HH:mm:ss'}}
```
Comment (owennw) [inline src/closedwindows/closedcard.html]: The spec says the date format should be 'dd MMM yyyy' I think.

File: src/closedwindows/closedcard.html [hunk 6]
```diff
@@ -0,0 +1,7 @@
...
+    </div>
+    <span class="closedstock" ng-repeat="stock in cwindow.stocks">
+        {{stock}} {{$last ? '' : ', '}}
```
Comment (owennw) [inline src/closedwindows/closedcard.html]: I imagine there's an unwanted space between the stock name and the comma here?

File: src/closedwindows/closedcard-controller.js (hunk 4 above)
Comment (bjedrzejewski) [inline src/closedwindows/closedcard-controller.js]: I don't understand. I thought this follows the pattern of every other controller.

File: src/closedwindows/closedcard.less [hunk 7]
```diff
@@ -0,0 +1,22 @@
+.closedtime {
+  font-size: 12px;
+  font-weight: 100;
```
Comment (owennw) [inline src/closedwindows/closedcard.less]: We have variables for font weights (this might be `@font-light`? not sure which though.)

File: src/closedwindows/closedcard.less [hunk 8]
```diff
@@ -0,0 +1,22 @@
...
+  font-size: 13px;
+}
+
+.closedcard {
+  height: 60px;
+  border-bottom: 1px solid @separator-colour;
+  overflow: hidden;
+  text-overflow: ellipsis;
+  padding: 5px;
+  cursor: default;
```
Comment (owennw) [inline src/closedwindows/closedcard.less]: Is the cursor line necessary?

File: src/closedwindows/closedcard.html (hunk 5 above)
Comment (bjedrzejewski) [inline src/closedwindows/closedcard.html]: True, just changing.

File: src/closedwindows/closedwindows-controller.js [hunk 9]
```diff
@@ -0,0 +1,19 @@
+(function() {
+    'use strict';
+
+    class ClosedWindowsCtrl {
+        constructor(storeService) {
+            this.storeService = storeService;
+            this.numbers = [1, 2, 3];
```
Comment (owennw) [inline src/closedwindows/closedwindows-controller.js]: Not entirely sure what this variable is for.

File: src/closedwindows/closedwindows-controller.js [hunk 10]
```diff
@@ -0,0 +1,19 @@
...
+            this.refreshClosedWindows();
+        }
+        refreshClosedWindows() {
+            this.storeService.refreshStore();
+            this.closedWindows = this.storeService.getPreviousClosedWindows().sort((a, b) => b.closed - a.closed);
+        }
+    }
+    ClosedWindowsCtrl.$inject = ['storeService'];
+
+    angular.module('openfin.closedwindows', [])
```
Comment (owennw) [inline src/closedwindows/closedwindows-controller.js]: Same comment about the injection here.

File: src/closedwindows/closedcard.html (hunk 6 above)
Comment (bjedrzejewski) [inline src/closedwindows/closedcard.html]: Well spotted.

File: src/closedwindows/closedcard.less (hunk 7 above)
Comment (bjedrzejewski) [inline src/closedwindows/closedcard.less]: Changed

File: src/sidebars/sidebar.html [hunk 11]
```diff
@@ -16,6 +16,20 @@
                 </div>
             </div>
         </div>
+        <div ng-controller="ClosedWindowsCtrl as closedWindowsCtrl">
+            <div class="favClosed">
+                <icon name="closed_tabs" ng-show="true" icon-click="closedTabsShow = true"
```
Comment (owennw) [inline src/sidebars/sidebar.html]: The `ng-show` is redundant here.

File: src/closedwindows/closedcard.less (hunk 8 above)
Comment (bjedrzejewski) [inline src/closedwindows/closedcard.less]: Yes, otherwise it keeps changing as you mouse over and looks blah

File: src/sidebars/sidebar.html [hunk 12]
```diff
@@ -16,6 +16,20 @@
...
+                      ng-click="$event.stopPropagation(); closedWindowsCtrl.refreshClosedWindows()"></icon>
```
Comment (owennw) [inline src/sidebars/sidebar.html]: I notice both `icon-click` and `ng-click` are being used here. Is there a reason for using both, or can they be combined?

File: src/closedwindows/closedwindows-controller.js (hunk 9 above)
Comment (bjedrzejewski) [inline src/closedwindows/closedwindows-controller.js]: Leftovers, removed.

File: src/closedwindows/closedwindows-controller.js [hunk 13]
```diff
@@ -0,0 +1,19 @@
...
+
+    class ClosedWindowsCtrl {
+        constructor(storeService) {
+            this.storeService = storeService;
+            this.numbers = [1, 2, 3];
+            this.refreshClosedWindows();
+        }
+        refreshClosedWindows() {
+            this.storeService.refreshStore();
+            this.closedWindows = this.storeService.getPreviousClosedWindows().sort((a, b) => b.closed - a.closed);
```
Comment (owennw) [inline src/closedwindows/closedwindows-controller.js]: I wonder whether the sorting should be performed in the HTML with angular. I think in general controllers are used for bindings, and the UI is responsible for knowing its own order. I think the native filter `orderBy` could be used to do it? I wonder what other people think though: @WPFerg ?

File: src/sidebars/sidebar.html [hunk 14]
```diff
@@ -16,6 +16,20 @@
...
+            </div>
+            <div ng-show="closedTabsShow" class="favClosedCover" ng-click="closedTabsShow = false">
+                <div class="closedSelection" ng-click="$event.stopPropagation()">
+                    <div ng-repeat="closedwindow in closedWindowsCtrl.closedWindows">
```
Comment (owennw) [inline src/sidebars/sidebar.html]: This is where that ordering filter mentioned above would be done.

File: src/sidebars/sidebar.less [hunk 15]
```diff
@@ -35,6 +35,33 @@
...
+}
+
+.closedSelection {
+  background: @light-background;
+  position: fixed;
+  bottom: 40px;
+  left: 240px;
+  width: 180px;
+  height: 300px;
+  box-shadow: 2px 2px 2px 2px @dark-background;
```
Comment (owennw) [inline src/sidebars/sidebar.less]: Is this what's doing the drop-shadow? That could be useful for our other windows...

File: src/sidebars/sidebar.html (hunk 11 above)
Comment (bjedrzejewski) [inline src/sidebars/sidebar.html]: This is WIP, will not show if no windows were closed.

File: src/closedwindows/closedcard-directive.js [hunk 16]
```diff
@@ -0,0 +1,14 @@
+(function() {
+    'use strict';
+
+    angular.module('openfin.closedcard')
+        .directive('closedcard', [function() {
```
Comment (owennw) [inline src/closedwindows/closedcard-directive.js]: Since you started this I changed all the `function()` over to `() =>`, so would be good to change this one too!

File: src/sidebars/sidebar.html (hunk 12 above)
Comment (bjedrzejewski) [inline src/sidebars/sidebar.html]: good point

File: src/closedwindows/closedwindows-controller.js (hunk 13 above)
Comment (bjedrzejewski) [inline src/closedwindows/closedwindows-controller.js]: I was wondering the same to be honest. Can refactor to get the orderBy

File: src/closedwindows/closedcard-controller.js [hunk 17]
```diff
@@ -0,0 +1,13 @@
...
+        .controller('ClosedcardCtrl', ClosedcardCtrl);
+}());
```
Comment (owennw) [inline src/closedwindows/closedcard-controller.js]: In fact.. is this controller used/necessary?

File: src/closedwindows/closedcard.less [hunk 18]
```diff
@@ -0,0 +1,22 @@
+.closedtime {
+  font-size: 12px;
+  font-weight: 100;
+  line-height: 1.4;
+}
+
+.closedstock {
```
Comment (owennw) [inline src/closedwindows/closedcard.less]: More camel-casing here too.

File: src/store-service.js [hunk 19]
```diff
@@ -104,6 +107,16 @@
                 .map((store) => store.id);
         }
 
+        getPreviousClosedWindowNames() {
```
Comment (owennw) [inline src/store-service.js]: I think this function is only called in here? If so it can be performed inline and taken off the interface.

File: src/closedwindows/closedcard.less (hunk 18 above)
Comment (bjedrzejewski) [inline src/closedwindows/closedcard.less]: Shall we standardised on `closed-time` etc. - using `-` instead?

File: src/store-service.js (hunk 19 above)
Comment (bjedrzejewski) [inline src/store-service.js]: I think it helps readability.

File: src/closedwindows/closedcard-directive.js (hunk 16 above)
Comment (bjedrzejewski) [inline src/closedwindows/closedcard-directive.js]: Why did you change all functions to () => ?

File: src/closedwindows/closedcard-controller.js (hunk 17 above)
Comment (bjedrzejewski) [inline src/closedwindows/closedcard-controller.js]: Yes, it is WIP.

File: src/app.js (hunk 1 above)
Comment (owennw) [inline src/app.js]: `minichart` is a bit of an oddball, perhaps it should be too. In any case, `currentWindow` and I think everything else should be.

File: src/closedwindows/closedcard-controller.js (hunk 4 above)
Comment (owennw) [inline src/closedwindows/closedcard-controller.js]: Talked about IRL.

File: src/closedwindows/closedcard-directive.js (hunk 16 above)
Comment (owennw) [inline src/closedwindows/closedcard-directive.js]: `function()` and `() =>` behave slightly different in the way they handle `this`; we'd had some bugs using the old pattern before so to remain consistent they were all changed. Hopefully it'll minimise the change of unexpected bugs around this in the future.

File: src/closedwindows/closedcard.less (hunk 18 above)
Comment (owennw) [inline src/closedwindows/closedcard.less]: Yes probably, though for all the others perhaps an issue should be raised.

**Comment by bjedrzejewski** [comment]
@owennw ok, changes made. Can you merge when you have time? I will finish some of the 'empty controller' things with the real functionality, no point committing just to revert now.

File: src/store-service.js [hunk 20]
```diff
@@ -104,6 +107,16 @@
...
+            return (this.storage || [])
+                .filter((store) => store.closed > 0)
+                .map((store) => store.id);
+        }
+
+        getPreviousClosedWindows() {
+            return this.getPreviousClosedWindowNames().map((name) => this.open(name).store);
```
Comment (owennw) [inline src/store-service.js]: This feels like a roundabout way to get the object you actually want. We know this object exists in the store ready, so I don't think we need to go to the trouble of creating a `StoreWrapper` by injecting the store object, just to fetch the object back off of it.

Comment (bjedrzejewski) [inline src/store-service.js]: I don't understand the comment, can we talk IRL?

**Comment by owennw** [comment]
I haven't finished reviewing it yet...

**Comment by bjedrzejewski** [comment]
Pull request #391 we can close this.
